src/NBitcoin/ChainedBlock.cs

```diff
@@ -449,13 +452,53 @@ public ChainedBlock GetAncestor(int height)
 
             ChainedBlock current = this;
 
-            while (true)
+            int currentHeight = this.Height;
+            while (currentHeight > height)
             {
-                if (current.Height == height)
-                    return current;
-
-                current = current.Previous;
+                int heightSkip = GetSkipHeight(currentHeight);
+                int heightSkipPrev = GetSkipHeight(currentHeight - 1);
+                if ((current.Skip != null) && ((heightSkip == height) || ((heightSkip > height) && !((heightSkipPrev < (heightSkip - 2)) && (heightSkipPrev >= height)))))
```

Comment:
there is too much magic on this line

consider the following implementation of the same thing
```
        public ChainedBlock GetAncestor(uint ancestorHeight)
        {
            if (ancestorHeight > this.Height)
                return null;

            ChainedBlock walk = this;
            while (walk.Height != ancestorHeight)
            {
                int heightSkip = this.GetSkipHeight(walk.Height);
                int heightSkipPrev = this.GetSkipHeight(walk.Height - 1);

                bool skipExists = walk.skip != null;
                bool skipIsTarget = heightSkip == ancestorHeight;
                bool skipAboveTarget = heightSkip > ancestorHeight;
                // Only follow skip if Previous.skip isn't better than skip.Previous.
                bool skipPreviousBetterThanPreviousSkip = !((heightSkipPrev < heightSkip - 2) && (heightSkipPrev >= ancestorHeight));
                bool useSkip = skipExists && (skipIsTarget || (skipAboveTarget && skipPreviousBetterThanPreviousSkip));
                walk = useSkip ? walk.skip : walk.Previous;
            }

            return walk;
        }
```

now i think this one is much more readable, but there is a cost for that and that is that in my version we evaluate all things even if it is not necessary

either that is a concern or not (my version has slightly worse performance)
if it is a concern, we could just use different structure of the code while keeping the performance 

@dangershony could you pls help us decide which way to go here? for me the one liner is completely unreadable, it is just to big to grasp, i have no idea what is it doing and why

Reply:
@bokobza and @noescape00 and @fassadlr can help here as well

Reply:
I agree @Aprogiena the readibility needs to be improved, how about this?

```
            if (walk.skip == null)
                return walk.Previous;
            else
            {
                if (heightSkip == ancestorHeight)
                    return walk.skip;

                bool skipAboveTarget = heightSkip > ancestorHeight;

                // Only follow skip if Previous.skip isn't better than skip.Previous.
                bool skipPreviousBetterThanPreviousSkip = !((heightSkipPrev < heightSkip - 2) && (heightSkipPrev >= ancestorHeight));

                if ((skipAboveTarget && skipPreviousBetterThanPreviousSkip))
                    return walk.skip;

                return walk.Previous;
            }
```

Reply:
when you return, you don't need else for the if

Reply:
Ah yes! Ive edited...

Reply:
that first return should actually be continue!?

Reply:
probably all returns should be assignments and continue in your code

Reply:
OK here is the whole thing in context of the While and his variable names (please verify):

```
            while (currentHeight > height)
            {
                int heightSkip = GetSkipHeight(currentHeight);
                int heightSkipPrev = GetSkipHeight(currentHeight - 1);

                if (current.Skip == null)
                {
                    current = current.Previous;
                    currentHeight--;

                    continue;
                }

                if (heightSkip == height)
                {
                    current = current.Skip;
                    currentHeight = heightSkip;

                    continue;
                }

                bool skipAboveTarget = heightSkip > ancestorHeight;
                bool skipPreviousBetterThanPreviousSkip = !((heightSkipPrev < heightSkip - 2) && (heightSkipPrev >= ancestorHeight));

                if (skipAboveTarget && skipPreviousBetterThanPreviousSkip)
                {
                    // Only follow Skip if Previous.Skip isn't better than Skip.Previous.
                    current = current.Skip;
                    currentHeight = heightSkip;
                }
                else
                {
                    current = current.Previous;
                    currentHeight--;
                }
            }
```

Reply:
hmm, that escalated to quite a long thing, my take 2 would be
```
        public ChainedBlock GetAncestor(uint ancestorHeight)
        {
            if (ancestorHeight > this.Height)
                return null;

            ChainedBlock walk = this;
            while (walk.Height != ancestorHeight)
            {
                int heightSkip = this.GetSkipHeight(walk.Height);
                int heightSkipPrev = this.GetSkipHeight(walk.Height - 1);

                if (walk.skip != null)
                {
                    bool skipIsTarget = heightSkip == ancestorHeight;
                    bool skipAboveTarget = heightSkip > ancestorHeight;
                    // Only follow skip if Previous.skip isn't better than skip.Previous.
                    bool skipPreviousBetterThanPreviousSkip = !((heightSkipPrev < heightSkip - 2) && (heightSkipPrev >= ancestorHeight));
                    bool useSkip = skipIsTarget || (skipAboveTarget && skipPreviousBetterThanPreviousSkip);
                    walk = useSkip ? walk.skip : walk.Previous;
                }
                else walk = walk.Previous;
            }

            return walk;
        }
```

Reply:
We shouldn't be scared of more if statements to make things more readable ;) Here's my last attempt:

```
            public ChainedBlock GetAncestor(uint ancestorHeight)
            {
                if (ancestorHeight > this.Height)
                    return null;

                ChainedBlock walk = this;
                while (walk.Height != ancestorHeight)
                {
                    int heightSkip = this.GetSkipHeight(walk.Height);
                    int heightSkipPrev = this.GetSkipHeight(walk.Height - 1);

                    if (walk.skip == null)
                    {
                        walk = walk.Previous;
                        continue;
                    }

                    if (heightSkip == ancestorHeight)
                    {
                        walk = walk.Skip;
                        continue;
                    }

                    bool skipAboveTarget = heightSkip > ancestorHeight;
                    // Only follow skip if Previous.skip isn't better than skip.Previous.
                    bool skipPreviousBetterThanPreviousSkip = !((heightSkipPrev < heightSkip - 2) && (heightSkipPrev >= ancestorHeight));
                    if (skipAboveTarget && skipPreviousBetterThanPreviousSkip)
                    {
                        walk = walk.Skip;
                        continue;
                    }

                    walk = walk.Previous;
                }

                return walk;
            }
```

Reply:
Ho wowoo I don't have an opinion here, too busy on other PRs so what ever works 😉

Reply:
Ideally there would be no comments, methods, variables or properties in ChainedBlock,cs that indicated the fact that it uses skip lists. A base class should be used for the chain and we should only see some high level methods calling it. We should be able to replace that base class without changing any other code in ChainedBlock.cs, If we can achieve that then we have properly separated the skip list implementation from the rest of the code.

Reply:
@fassadlr i don't necessarily agree that longer code with more IFs is more readable, sometimes it is, sometimes it is not

Reply:
but we do have couple of options here 
so let's leave it to @mikedennis which version he picks

Reply:
@quantumagi thanks for the input, but we it is kind of offtopic in this context, which is solely about implementing this one function, what you are talking about is higher level design